File: 03-bookingservice/Domain/ActorType.cs
namespace bookingservice.Domain;

public enum ActorType
{
    Customer,
    Corp
}

public static class ActorTypeExtensions
{
    public static ActorType FromContext(string? actorTypeHeader, string? realmHeader)
    {
        if (string.Equals(Trim(actorTypeHeader), "corp", StringComparison.OrdinalIgnoreCase))
        {
            return ActorType.Corp;
        }

        if (string.Equals(Trim(actorTypeHeader), "customer", StringComparison.OrdinalIgnoreCase))
        {
            return ActorType.Customer;
        }

        var realm = Trim(realmHeader);
        if (string.Equals(realm, "corp", StringComparison.OrdinalIgnoreCase) ||
            string.Equals(realm, "voyagevibes-corp", StringComparison.OrdinalIgnoreCase))
        {
            return ActorType.Corp;
        }

        return ActorType.Customer;
    }

    public static string HeaderValue(this ActorType actorType) => actorType == ActorType.Corp ? "corp" : "customer";

    private static string Trim(string? value) => value?.Trim() ?? string.Empty;
}

File: 02-flightservice/src/main/java/com/cloudxplorer/flightservice/domain/ActorType.java
package com.cloudxplorer.flightservice.domain;

public enum ActorType {
  CUSTOMER,
  CORP;

  public static ActorType fromHeader(String value) {
    return fromContext(value, null);
  }

  public static ActorType fromContext(String actorTypeHeader, String realmHeader) {
    if ("corp".equalsIgnoreCase(trim(actorTypeHeader))) {
      return CORP;
    }
    if ("customer".equalsIgnoreCase(trim(actorTypeHeader))) {
      return CUSTOMER;
    }

    String realm = trim(realmHeader);
    if ("corp".equalsIgnoreCase(realm) || "voyagevibes-corp".equalsIgnoreCase(realm)) {
      return CORP;
    }
    return CUSTOMER;
  }

  public String headerValue() {
    return name().toLowerCase();
  }

  private static String trim(String value) {
    return value == null ? "" : value.trim();
  }
}
